torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +127,138 @@ def _warn_on_overridden_mixed_precision(
+    """
+    topo_sorted_named_modules = _get_topo_sorted_named_modules(root_module)
+    reverse_topo_sorted_named_modules = reversed(topo_sorted_named_modules)
+    visited_modules: Set[nn.Module] = set()
+    for module_name, module in reverse_topo_sorted_named_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_managed_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            frozen_param_numel = 0
+            nonfrozen_param_fqns: List[str] = []
+            nonfrozen_param_numel = 0
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                    nonfrozen_param_numel += param.numel()
+                else:
+                    frozen_param_fqns.append(fqn)
+                    frozen_param_numel += param.numel()
+            if len(frozen_param_fqns) > 0 and len(nonfrozen_param_fqns) > 0:
+                msg = f"{module_name} has both parameters with requires_grad=True and False."
+                if use_orig_params:
+                    total_param_numel = frozen_param_numel + nonfrozen_param_numel
+                    msg += (
+                        " We do not recommend wrapping such modules since "
+                        "the gradient memory usage will be higher than expected "
+                        f"({total_param_numel} numel instead of {nonfrozen_param_numel} numel "
+                        "before sharding via reduce-scatter). "
+                    )
+                else:
+                    msg += " FSDP does not support wrapping such modules when use_orig_params=False. "
+                msg += "If possible, wrap the frozen parameters with FSDP separately.\n"
+                msg += (
+                    f"The following parameters have requires_grad=True:\n{nonfrozen_param_fqns}\n"
+                    f"The following parameters have requires_grad=False:\n{frozen_param_fqns}"
+                )
+                if use_orig_params:
+                    warnings.warn(msg)
+                else:
+                    raise ValueError(msg)
+
+
+def _get_topo_sorted_named_modules(root_module: nn.Module) -> List[nn.Module]:
+    """
+    We use a stack-based DFS instead of ``root_module.named_modules()`` to get
+    the topological order since reversing the DFS order gives the modules in
+    registration order at each level in the module tree (as opposed to in
+    reverse), which allows us to error/warn on the first registered module that
+    violates the condition.
+
+    For example, consider the following module structure:
+        M(
+          S1(),
+          S2(
+            SS1(),
+            SS2(),
+          ),
+          S3(),
+        )
+    The reverse DFS order is [S1, SS1, SS2, S2, S3, M], while the reverse
+    ``named_modules()`` order is [S3, SS2, SS1, S2, S1, M].
```

Comment:
So, if I'm right, and this is just a postorder traversal, you should just call it post-order traversal, not topo-sort, because topo-sort is underspecified: `named_modules` order is also a valid topsort, it's just not the one you want. Since it's important that you have a postorder here, should call it that.

Reply:
Makes sense!